### PR TITLE
Remove trivia tokens

### PR DESCRIPTION
**Pull request by matklad**
r? @ghost

### CODE REVIEW

**Comment by matklad** [comment]
Builds on the #76166

This doesn't actually remove associated enum variants just yet, but, surprisingly, just never constructiong those variants works without any additional tweaks anywhere. 

cc @petrochenkov

**Comment by matklad** [comment]
I think the end result here would be something like

```rust
crate fn tokenize_into_token_trees(
    sess: &'a ParseSess,
    source_file: Lrc<rustc_span::SourceFile>,
    override_span: Option<Span>,
) -> (PResult<'a, TokenStream>, Vec<UnmatchedBrace>)
```

Ie, `StringReader` and `TokenTreesReader` would become, at minimum, private, but may actually get completely dissolved

File: compiler/rustc_parse/src/lexer/mod.rs [hunk 1]
```diff
@@ -140,8 +143,8 @@ impl<'a> StringReader<'a> {
     /// Turns simple `rustc_lexer::TokenKind` enum into a rich
     /// `librustc_ast::TokenKind`. This turns strings into interned
     /// symbols and runs additional validation.
-    fn cook_lexer_token(&self, token: rustc_lexer::TokenKind, start: BytePos) -> TokenKind {
-        match token {
+    fn cook_lexer_token(&self, token: rustc_lexer::TokenKind, start: BytePos) -> Option<TokenKind> {
+        let res = match token {
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/lexer/mod.rs:147]:
```suggestion
        Some(match token {
```

**Comment by matklad** [comment]
Ready for review @petrochenkov! 

I've decided to leave overall code shape changes (fully privatising StringReader) to a follow up PR, to separate pure refactoring from changes to core data structures.

File: compiler/rustc_parse/src/lexer/mod.rs (hunk 1 above)
Comment (matklad) [reply compiler/rustc_parse/src/lexer/mod.rs]: Urgh, thanks, sorry for not remembering that this is the preferred style in rustc!

**Comment by matklad** [comment]
Wait, tests fail locally...

File: compiler/rustc_parse/src/lexer/tokentrees.rs [hunk 2]
```diff
@@ -264,27 +262,19 @@ impl<'a> TokenTreesReader<'a> {
             }
             _ => {
                 let tt = TokenTree::Token(self.token.take());
-                self.real_token();
-                let is_joint = self.joint_to_prev == Joint && self.token.is_op();
-                Ok((tt, if is_joint { Joint } else { NonJoint }))
+                let mut is_joint = self.bump();
+                if !self.token.is_op() {
+                    is_joint = NonJoint;
+                }
```
Comment (matklad) [inline compiler/rustc_parse/src/lexer/tokentrees.rs:268]: I originally forgot this condition, and that failed some proc-macro tests. This is a bit weird -- I would expect this jointness censoring to happen at the proc_macro_srv layer, and not here.

**Comment by matklad** [comment]
Ok, should be good to go now!

File: compiler/rustc_parse/src/lexer/tokentrees.rs (hunk 2 above)
Comment (petrochenkov) [reply compiler/rustc_parse/src/lexer/tokentrees.rs:268]: Yes, this check should be moved to proc macro server.
Also lexer should produce the jointness flag for the "delimited group" token trees as well (the match branches above).

Comment (matklad) [reply compiler/rustc_parse/src/lexer/tokentrees.rs:268]: Should I do this in this PR, or a separate one?

I've also rebased #75528 on top of this PR, seems to work

https://github.com/matklad/rust/commit/21bf6ce57289a50be8bced3abe44e23ca9ceeb56

Comment (petrochenkov) [reply compiler/rustc_parse/src/lexer/tokentrees.rs:268]: >Should I do this in this PR, or a separate one?

Whichever is more convenient.

**Pull request merged**